pkgs/development/libraries/protobuf/generic-v3.nix

```diff
@@ -1,60 +1,68 @@
-{ stdenv
-, fetchFromGitHub
-, autoreconfHook, zlib, gmock, buildPackages
+{ stdenv, fetchFromGitHub, lib
+, zlib, cmake, gmock
 , version, sha256
 , ...
 }:
-
-let
-mkProtobufDerivation = buildProtobuf: stdenv: stdenv.mkDerivation {
+stdenv.mkDerivation rec {
+  # make sure you test also -A pythonPackages.protobuf
   pname = "protobuf";
   inherit version;
-
-  # make sure you test also -A pythonPackages.protobuf
   src = fetchFromGitHub {
     owner = "protocolbuffers";
     repo = "protobuf";
     rev = "v${version}";
     inherit sha256;
   };
 
-  postPatch = ''
-    rm -rf gmock
-    cp -r ${gmock.src}/googlemock gmock
-    cp -r ${gmock.src}/googletest googletest
-    chmod -R a+w gmock
-    chmod -R a+w googletest
-    ln -s ../googletest gmock/gtest
-  '' + stdenv.lib.optionalString stdenv.isDarwin ''
-    substituteInPlace src/google/protobuf/testing/googletest.cc \
-      --replace 'tmpnam(b)' '"'$TMPDIR'/foo"'
-  '';
+  patches =
+    [ ./protobuf-cmake.patch ]
+    # In version 3.6.X,
+    # CMake doesn't generate the version.rc file from version.rc.in.
+    ++ lib.optional
+       (lib.versionAtLeast version "3.6" && lib.versionOlder version "3.7")
+       ./protobuf-3.6-cmake-versionrc.patch;
 
-  nativeBuildInputs = [ autoreconfHook buildPackages.which buildPackages.stdenv.cc buildProtobuf ];
+  postPatch = 
+    (if lib.versionAtLeast version "3.6.0"
+     then ''
+       rm -rf third_party/googletest
+       cp -a ${gmock.src} third_party/googletest
+       chmod -R a+w third_party/googletest
+       ''
+     else ''
+     cp -a ${gmock.src}/googlemock gmock
+     cp -a ${gmock.src}/googletest googletest
+     chmod -R a+w gmock
+     chmod -R a+w googletest
+     ln -s ../googletest gmock/gtest
+     '')
```

Comment:
Can set `fetchSubmodules = true` in `fetchFromGithub` (don't forget to update hashes).

Reply:
That is an approach I could take. I borrowed this approach from the existing Autoconf build.

Reply:
Let's try that. Also remove the darwin hack below, we'll figure something out if it's still needed.

Reply:
(Don't forget to drop `gmock` argument)